gui/src/shared/relay-location-builder.ts

```diff
@@ -0,0 +1,50 @@
+import { Constraint, LiftedConstraint, RelayLocation } from './daemon-rpc-types';
+
+export interface ILocationBuilder<Self> {
+  country(country: string): Self;
+  city(country: string, city: string): Self;
+  hostname(country: string, city: string, hostname: string): Self;
+  any(): Self;
+  fromRaw(location: LiftedConstraint<RelayLocation>): Self;
+}
+
+export default function makeLocationBuilder<T>(
+  context: T,
+  receiver: (constraint: Constraint<RelayLocation>) => void,
+): ILocationBuilder<T> {
+  return {
+    country: (country: string) => {
+      receiver({ only: { country } });
+      return context;
+    },
+    city: (country: string, city: string) => {
+      receiver({ only: { city: [country, city] } });
+      return context;
+    },
+    hostname: (country: string, city: string, hostname: string) => {
+      receiver({ only: { hostname: [country, city, hostname] } });
+      return context;
+    },
+    any: () => {
+      receiver('any');
+      return context;
+    },
+    fromRaw(location: LiftedConstraint<RelayLocation>) {
+      if (location === 'any') {
```

Comment:
Can't this be simplified using the new type information? I.e., could it be simplified to do something like:

```
if (location === 'any') {
    return this.any();
} else {
    receiver({ only: location });
    return context;
}
```

Reply:
Ah sure! Done!